### PR TITLE
[BEAM-7502] Create ParDo Python Load Test Jenkins job

### PR DESCRIPTION
**Pull request by kamilwu**
Based on the following proposal: https://s.apache.org/load-test-basic-operations

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

**Comment by kamilwu** [comment]
@lgajowy Please take a look.
I had to increase the number of workers from 5 (as in the proposal) to 10, because the job had been running for longer than 2 hours. As a result, it was always aborted.

File: .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                        publish_to_big_query : true,
+                        metrics_dataset      : datasetName,
+                        metrics_table        : 'python_dataflow_batch_pardo_1',
+                        input_options        : '\'{' +
+                                '"num_records": 20000000,' +
+                                '"key_size": 10,' +
+                                '"value_size": 90}\'',
+                        iterations           : 10,
+                        number_of_counter_operations: 0,
+                        number_of_counters   : 1,
```
Comment (lgajowy) [inline .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy]: Why is there 1 counter? According to the proposal the goal of this test is to check what is the inter operation overhead (not metrics). So there should be no counters. Counters are needed in tests #3 and #4 where we examine metrics overhead

File: .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                        num_workers          : 10,
```
Comment (lgajowy) [inline .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy]: Why are we using 10 workers here?

File: .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                        autoscaling_algorithm: 'NONE',
+                ]
+        ],
+        [
+                title        : 'ParDo Python Load test: 2GB 100 byte records 200 times',
+                itClass      : 'apache_beam.testing.load_tests.pardo_test:ParDoTest.testParDo',
+                runner       : CommonTestProperties.Runner.DATAFLOW,
+                sdk          : CommonTestProperties.SDK.PYTHON,
+                jobProperties: [
```
Comment (lgajowy) [inline .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy]: (same batch of questions as to test scenario no. 1)

**Comment by kamilwu** [comment]
Run Seed Job

File: .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                        iterations           : 10,
+                        number_of_counter_operations: 10,
+                        number_of_counters   : 1,
+                        num_workers          : 5,
+                        autoscaling_algorithm: 'NONE',
+                ]
+        ],
+]}
+
+def loadTestConfigurationManyCounters = { datasetName -> [
```
Comment (lgajowy) [inline .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy]: I don't think splitting the job to 2 jobs is the solution here. It is very weird that it takes so much time (above 2 hours) to run, whereas java needs 28 minutes to run all the tests (as we discussed offline). Could you investigate this a little bit more? Maybe the pipeline shape is not as we think it is?

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

File: .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy (hunk 1 above)
Comment (kamilwu) [reply .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy]: The `number_of_counter_operations` is zero, so there is actually no metrics overhead.
But I guess this 1 counter can be misleading, so I'll change it.

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

File: .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+def loadTestConfigurations = { datasetName -> [
+        [
+                title        : 'ParDo Python Load test: 2GB 100 byte records 10 times',
+                itClass      : 'apache_beam.testing.load_tests.pardo_test:ParDoTest.testParDo',
+                runner       : CommonTestProperties.Runner.DATAFLOW,
+                sdk          : CommonTestProperties.SDK.PYTHON,
+                jobProperties: [
+                        job_name             : 'load-tests-python-dataflow-batch-pardo-1-' + now,
+                        project              : 'apache-beam-testing',
+                        temp_location        : 'gs://temp-storage-for-perf-tests/smoketests',
```
[the remark this answers is not in the document]
Comment (kamilwu) [reply .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy]: It's correct, but 'gs://temp-storage-for-perf-tests/loadtests' seems to be used more often in load tests.

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

File: .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy (hunk 5 above)
[the remark this answers is not in the document]
Comment (lgajowy) [reply .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy]: Could you explain why is this a problem? Did you experience any issues when using /loadtests?

[the remark this answers is not in the document]
Comment (kamilwu) [reply .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy]: I think you misunderstood. It's not a problem. `/smoketests` was there accidentally, and I've already replaced it to `/loadtests`

[the remark this answers is not in the document]
Comment (lgajowy) [reply .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy]: Right, we're seeing stale code in the thread. Thanks! :)

**Comment by kamilwu** [comment]
Run Seed Job

File: .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy (hunk 2 above)
Comment (kamilwu) [reply .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy]: I've already changed it to 5.

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

File: .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy (hunk 4 above)
Comment (kamilwu) [reply .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy]: I have double-checked the code and it seems fine. The problem is poor performance of metrics operations in Python — test case with 10 iterations and 100 counter operations needed almost 3 hours to complete.

The solution is to lower the number of iterations to 1, which makes the test way faster. Job split will be also unnecessary.

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

File: .test-infra/jenkins/job_LoadTests_ParDo_Java.groovy [hunk 6]
```diff
@@ -97,7 +97,7 @@ def commonLoadTestConfig = { jobType, isStreaming, datasetName ->
                                                       "valueSizeBytes": 90
                                                     }
                                                """.trim().replaceAll("\\s", ""),
-                            iterations          : 10,
+                            iterations          : 1,
```
Comment (lgajowy) [inline .test-infra/jenkins/job_LoadTests_ParDo_Java.groovy]: Agree that this should be done in java too. Nit before merging - while you'll be working on the branch history - could you introduce the java change in separate commit?

Comment (kamilwu) [reply .test-infra/jenkins/job_LoadTests_ParDo_Java.groovy]: Alright

**Comment by kamilwu** [comment]
@lgajowy It's done, commits are ready

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

**Pull request merged**